File: extensions/messenger/src/main/java/app/morphe/extension/messenger/metaai/RemoveMetaAIPatch.java
/*
 * Forked from:
 * https://github.com/ReVanced/revanced-patches/blob/6b06b9d1328b971a06d10b4247f4c10f050e4f61/extensions/messenger/src/main/java/app/revanced/extension/messenger/metaai/RemoveMetaAIPatch.java
 */

package app.morphe.extension.messenger.metaai;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import app.morphe.extension.shared.Logger;

@SuppressWarnings("unused")
public class RemoveMetaAIPatch {
    private static final Set<Long> loggedIDs = Collections.synchronizedSet(new HashSet<>());

    public static boolean overrideBooleanFlag(long id, boolean value) {
        try {
            if (Long.toString(id).startsWith("REPLACED_BY_PATCH")) {
                if (loggedIDs.add(id)) {
                    Logger.printInfo(() -> "Overriding " + id + " from " + value + " to false");
                }
                return false;
            }
        } catch (Exception ex) {
            Logger.printException(() -> "overrideBooleanFlag failure", ex);
        }

        return value;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/messenger/metaai/RemoveMetaAIPatch.kt
/*
 * Forked from:
 * https://github.com/ReVanced/revanced-patches/blob/6b06b9d1328b971a06d10b4247f4c10f050e4f61/patches/src/main/kotlin/app/revanced/patches/messenger/metaai/RemoveMetaAIPatch.kt
 */

package app.morphe.patches.messenger.metaai

import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.extensions.InstructionExtensions.replaceInstruction
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.messenger.misc.extension.sharedExtensionPatch
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
import com.android.tools.smali.dexlib2.iface.instruction.WideLiteralInstruction

internal const val EXTENSION_CLASS_DESCRIPTOR = "Lapp/morphe/extension/messenger/metaai/RemoveMetaAIPatch;"
internal const val EXTENSION_METHOD_NAME = "overrideBooleanFlag"

@Suppress("unused")
val removeMetaAIPatch = bytecodePatch(
    name = "Remove Meta AI",
    description = "Removes UI elements related to Meta AI."
) {
    compatibleWith("com.facebook.orca")

    dependsOn(sharedExtensionPatch)

    execute {
        GetMobileConfigBoolFingerprint.method.apply {
            val returnIndex = GetMobileConfigBoolFingerprint.instructionMatches.first().index
            val returnRegister = getInstruction<OneRegisterInstruction>(returnIndex).registerA

            addInstructions(
                returnIndex,
                """
                    invoke-static { p1, p2, v$returnRegister }, $EXTENSION_CLASS_DESCRIPTOR->$EXTENSION_METHOD_NAME(JZ)Z
                    move-result v$returnRegister
                """
            )
        }

        // Extract the common starting digits of Meta AI flag IDs from a flag found in code.
        val relevantDigits = with(MetaAIKillSwitchCheckFingerprint) {
            method.getInstruction<WideLiteralInstruction>(instructionMatches.first().index).wideLiteral
        }.toString().substring(0, 7)

        // Replace placeholder in the extension method.
        with(ExtensionMethodFingerprint) {
            method.replaceInstruction(
                stringMatches.first().index,
                """
                    const-string v1, "$relevantDigits"
                """
            )
        }
    }
}
